src/js/timer.js

```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+
+    uBlock Origin - a browser extension to block requests.
+    Copyright (C) 2014-present Raymond Hill
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see {http://www.gnu.org/licenses/}.
+
+    Home: https://github.com/gorhill/uBlock
+*/
+
+/* globals requestIdleCallback, cancelIdleCallback, performance */
+
+'use strict';
+
+/******************************************************************************/
+
+const timer = {};
+
+timer.requestIdleCallback = typeof requestIdleCallback !== 'undefined' ?  requestIdleCallback : null;
+timer.cancelIdleCallback = typeof cancelIdleCallback !== 'undefined' ?  cancelIdleCallback : null;
+
+/******************************************************************************/
+
+const needsPolyfill = timer.requestIdleCallback === null;
+
+if ( needsPolyfill ) {
+    // https://github.com/behnammodi/polyfill/blob/1a5965edc0e2eaf8e6d87902cc719462e2a889fb/window.polyfill.js
```

Comment:
The code below is from the polyfill linked from here: https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback

Reply:
I prefer to keep the implementation simple -- I had a look at this polyfill back when I created `globals` but decided against it because there was no use for uBO of more complicated than what is there, the use-case in uBO is simply to defer some operations so as to not delay readiness at launch time. I prefer your idea of a `common` variable allowing access to common utility functions throughout different parts of the API. For instance `requestIdleCallback()` could simply be a custom-named method using `requestIdleCallback()` underneath, but without having all the semantic of `requestIdleCallback()`.

Reply:
What about a simple API like this?

```js
let id = timer.queue(myFunc);
timer.drop(id);
```

`timer.drop()` is not even necessary for now.

`static-net-filtering.js` does pass a `timeout` value so I'm wondering if that should be supported.

Reply:
We could use a default `timeout` internally without requiring it to be part of the API -- the MDN doc strongly suggests the use of a timeout to avoid cases where the callback may never be called.

Reply:
I've renamed the file to `tasks.js` and it's a lot simpler now.

Reply:
> I prefer your idea of a `common` variable allowing access to common utility functions throughout different parts of the API.

We could call this file `common.js` if you prefer it. Afterwards, when there's a solution for `location`, that could go in this file too.